api/main.go

```diff
@@ -96,3 +94,49 @@ func getSegmentAndIds(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 }
+
+// keycloak authentication function that wraps handlers needing keycloak auth
+func keycloakAuth(fn func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+
+		authHeader := strings.Fields(r.Header.Get("Authorization"))
+
+		if len(authHeader) < 1 {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, "Invalid authorization header")
+			return
+		}
+
+		token := authHeader[1]
+		kcClientID := os.Getenv("KC_CLIENT_ID")
+		kcClientSecret := os.Getenv("KC_CLIENT_SECRET")
+		kcRealm := os.Getenv("KC_REALM")
+		kcURL := os.Getenv("KC_URL")
+		kcClient := gocloak.NewClient((kcURL))
+
+		ctx := context.Background()
+		_, err := kcClient.LoginClient(ctx, kcClientID, kcClientSecret, kcRealm)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			fmt.Fprintf(w, "Keycloak Login failed:"+err.Error())
+			return
+		}
+
+		rptResult, err := kcClient.RetrospectToken(ctx, token, kcClientID, kcClientSecret, kcRealm)
```

Comment:
retrosepct token? does this call intro spection endpoint? Since you aren't revoking tokens you wouldn't need to introspect

Reply:
Updated to GetUserInfo